test/extensions/transport_sockets/tls/cert_validator/default_validator_integration_test.cc

```diff
@@ -0,0 +1,98 @@
+#include "default_validator_integration_test.h"
+
+#include <memory>
+
+#include "source/extensions/transport_sockets/tls/context_manager_impl.h"
+
+#include "test/integration/integration.h"
+
+#include "gtest/gtest.h"
+
+namespace Envoy {
+namespace Ssl {
+
+void SslCertValidatorIntegrationTest::initialize() {
+  HttpIntegrationTest::initialize();
+  context_manager_ =
+      std::make_unique<Extensions::TransportSockets::Tls::ContextManagerImpl>(timeSystem());
+  registerTestServerPorts({"http"});
+}
+
+void SslCertValidatorIntegrationTest::TearDown() {
+  HttpIntegrationTest::cleanupUpstreamAndDownstream();
+  codec_client_.reset();
+  context_manager_.reset();
+}
+
+Network::ClientConnectionPtr
+SslCertValidatorIntegrationTest::makeSslClientConnection(const ClientSslTransportOptions& options) {
+  ClientSslTransportOptions modified_options{options};
+  modified_options.setTlsVersion(tls_version_);
+  modified_options.setClientWithIntermediateCert(true);
+
+  Network::Address::InstanceConstSharedPtr address = getSslAddress(version_, lookupPort("http"));
+  auto client_transport_socket_factory_ptr =
+      createClientSslTransportSocketFactory(modified_options, *context_manager_, *api_);
+  return dispatcher_->createClientConnection(
+      address, Network::Address::InstanceConstSharedPtr(),
+      client_transport_socket_factory_ptr->createTransportSocket({}), nullptr);
+}
+
+void SslCertValidatorIntegrationTest::checkVerifyErrorCouter(uint64_t value) {
+  Stats::CounterSharedPtr counter =
+      test_server_->counter(listenerStatPrefix("ssl.fail_verify_error"));
+  EXPECT_EQ(value, counter->value());
```

Comment:
Integration tests end up flaky when you use this pattern. Usually it is better to do `test_server_->waitForCounterGe()` or similar. When checking that this counter is zero, first do a `waitForCounterGe()` on a metric that will increment in the test case, and after that, yhou can `EXPECT_EQ(0, the-zero-counter)`; that ordering should keep the test from being flaky.

Reply:
ok.. most of this is a copy from spifee-validator integration test